File: app/src/main/java/com/jin/nfcwrite/ColorPalette.kt
package com.jin.nfcwrite

import android.graphics.Color

/**
 * ColorPalette
 * -----------
 * DMN036EW 배지가 표시할 수 있는 6가지 색상을 정의하는 객체입니다.
 *
 * "object"로 선언하면 코틀린에서는 이 클래스의 인스턴스가 앱 전체에서
 * 딱 하나만 존재하게 됩니다 (싱글턴 패턴). new ColorPalette() 처럼
 * 따로 만들 필요 없이 ColorPalette.nearestColor(...) 형태로 바로 씁니다.
 *
 * TODO: 아래 rgb 값과 code 값은 임시로 넣어둔 것입니다.
 * 제조사(GooDisplay) 회신에서 "정확한 6색 RGB 기준값"과
 * "각 색이 몇 번 비트코드에 대응하는지"를 알려주면 이 값들을 교체해야 합니다.
 */
object ColorPalette {

    /**
     * 팔레트에 들어가는 색상 하나를 표현하는 데이터 클래스.
     * - name: 사람이 알아보기 위한 이름 (디버깅용)
     * - rgb: 안드로이드에서 쓰는 정수형 색상값 (0xAARRGGBB 형태)
     * - code: 배지 칩에 보낼 때 쓰는 비트 코드 (예: 000=검정, 001=흰색 등)
     */
    data class PaletteColor(val name: String, val rgb: Int, val code: Int)

    // 제조사(GooDisplay) 회신 기준 정확한 값으로 교체 (2026-07-20)
    // 참고: 0100(4)은 정의되지 않은 코드라 사용하지 않음(RFU로 추정)
    val SPECTRA6 = listOf(
        PaletteColor("Black",  Color.parseColor("#000000"), 0b0000), // 0
        PaletteColor("White",  Color.parseColor("#FFFFFF"), 0b0001), // 1
        PaletteColor("Yellow", Color.parseColor("#FFFF00"), 0b0010), // 2
        PaletteColor("Red",    Color.parseColor("#FF0000"), 0b0011), // 3
        PaletteColor("Blue",   Color.parseColor("#0000FF"), 0b0101), // 5
        PaletteColor("Green",  Color.parseColor("#00FF00"), 0b0110)  // 6
    )

    /**
     * 임의의 픽셀 색상(pixel)을 받아서, 팔레트 안에서 가장 비슷한 색을 찾아 반환합니다.
     *
     * 원리: "색상 거리"라는 개념을 씁니다.
     * 두 색의 R값 차이, G값 차이, B값 차이를 각각 제곱해서 더하면
     * 두 색이 얼마나 "떨어져 있는지"를 숫자로 나타낼 수 있습니다.
     * (이건 사실 3차원 공간에서 두 점 사이의 거리를 구하는 것과 똑같은 원리입니다.
     *  R,G,B를 x,y,z 좌표라고 생각하면 됩니다.)
     *
     * 이 거리가 가장 작은(=가장 비슷한) 팔레트 색을 찾아서 리턴합니다.
     */
    fun nearestColor(pixel: Int, palette: List<PaletteColor> = SPECTRA6): PaletteColor {
        // 원본 픽셀에서 R, G, B 값을 각각 추출 (0~255 범위)
        val r = Color.red(pixel)
        val g = Color.green(pixel)
        val b = Color.blue(pixel)

        var best = palette[0]           // 지금까지 찾은 "가장 가까운 색" (일단 첫 번째로 초기화)
        var bestDist = Int.MAX_VALUE    // 지금까지 찾은 "가장 작은 거리" (일단 매우 큰 값으로 초기화)

        // 팔레트에 있는 색상 하나하나를 순회하면서 거리를 계산
        for (p in palette) {
            val pr = Color.red(p.rgb)
            val pg = Color.green(p.rgb)
            val pb = Color.blue(p.rgb)

            // 유클리드 거리의 제곱 (제곱근은 생략 - 대소 비교만 할 거라 필요 없음)
            val dist = (r - pr) * (r - pr) + (g - pg) * (g - pg) + (b - pb) * (b - pb)

            // 지금까지 찾은 것보다 더 가까우면 갱신
            if (dist < bestDist) {
                bestDist = dist
                best = p
            }
        }
        return best
    }
}

File: pc-writer/color_palette.py
"""
color_palette.py
-----------------
DMN036EW 배지가 표시할 수 있는 6가지 색상을 정의합니다.
안드로이드 앱의 ColorPalette.kt와 완전히 동일한 값을 씁니다
(제조사 GooDisplay 회신 기준으로 확정된 값, 2026-07-20).

색상 거리 계산 방식은 제조사 공식 PC 프로그램(app.asar)을 직접
역분석해서 확인한 것을 그대로 사용합니다: 단순 RGB 유클리드 거리가
아니라, 사람 눈이 초록색 차이에 훨씬 민감하고 파란색 차이에는 둔감한
것을 반영한 가중치(Rec.709 휘도 계수: R=0.2126, G=0.7152, B=0.0722)를
곱한 거리입니다. 이게 애매한 회색/밝은 톤이 엉뚱한 색으로 튀는 문제를
크게 줄여준다는 것을 제조사 코드에서 확인했습니다.
"""

from dataclasses import dataclass
from typing import Tuple, List


@dataclass(frozen=True)
class PaletteColor:
    name: str
    rgb: Tuple[int, int, int]  # (R, G, B), 0~255
    code: int  # 배지 칩에 보낼 때 쓰는 4비트 코드


# 참고: 0b0100(4)는 정의되지 않은 코드라 사용하지 않음 (RFU로 추정)
SPECTRA6: List[PaletteColor] = [
    PaletteColor("Black", (0x00, 0x00, 0x00), 0b0000),   # 0
    PaletteColor("White", (0xFF, 0xFF, 0xFF), 0b0001),   # 1
    PaletteColor("Yellow", (0xFF, 0xFF, 0x00), 0b0010),  # 2
    PaletteColor("Red", (0xFF, 0x00, 0x00), 0b0011),     # 3
    PaletteColor("Blue", (0x00, 0x00, 0xFF), 0b0101),    # 5
    PaletteColor("Green", (0x00, 0xFF, 0x00), 0b0110),   # 6
]

# 제조사 코드에서 그대로 가져온 색상 거리 가중치 (Rec.709 휘도 계수)
WEIGHT_R = 0.2126
WEIGHT_G = 0.7152
WEIGHT_B = 0.0722

# numpy 기반 빠른 최근접색 계산에 쓰기 편하도록 배열 형태로도 준비
import numpy as np  # noqa: E402

PALETTE_RGB_ARRAY = np.array([c.rgb for c in SPECTRA6], dtype=np.int32)  # (6, 3)
PALETTE_CODE_ARRAY = np.array([c.code for c in SPECTRA6], dtype=np.uint8)  # (6,)
_DIST_WEIGHTS = np.array([WEIGHT_R, WEIGHT_G, WEIGHT_B], dtype=np.float64)  # (3,)


def nearest_color_index(pixel_rgb: Tuple[int, int, int]) -> int:
    """단일 픽셀(R,G,B)에 대해 팔레트 안에서 가장 가까운 색의 인덱스를 돌려줍니다."""
    r, g, b = pixel_rgb
    best_idx = 0
    best_dist = None
    for i, c in enumerate(SPECTRA6):
        pr, pg, pb = c.rgb
        dr, dg, db = r - pr, g - pg, b - pb
        dist = WEIGHT_R * dr * dr + WEIGHT_G * dg * dg + WEIGHT_B * db * db
        if best_dist is None or dist < best_dist:
            best_dist = dist
            best_idx = i
    return best_idx


def nearest_color_indices_bulk(pixels: "np.ndarray") -> "np.ndarray":
    """
    (H, W, 3) 모양의 numpy 배열을 받아서, 각 픽셀마다 팔레트에서 가장 가까운
    색의 인덱스를 담은 (H, W) 배열을 돌려줍니다. 픽셀 하나씩 파이썬 반복문을
    도는 것보다 훨씬 빠릅니다 (numpy 브로드캐스팅 사용).
    """
    # pixels: (H, W, 3), PALETTE_RGB_ARRAY: (6, 3)
    # 각 픽셀 대 각 팔레트색 사이의 거리를 한 번에 계산: (H, W, 6)
    diff = pixels[:, :, None, :].astype(np.float64) - PALETTE_RGB_ARRAY[None, None, :, :]
    dist_sq = np.sum(diff * diff * _DIST_WEIGHTS[None, None, None, :], axis=3)  # (H, W, 6)
    return np.argmin(dist_sq, axis=2)  # (H, W)
